### PR TITLE
Reland: [libc++][format] P2637R3: Member visit (std::basic_format_arg) #76449

### PR DESCRIPTION
**Pull request by H-G-Hristov**
Deleted the offending test case.

`libcxx/test/std/utilities/format/format.arguments/format.arg/visit.return_type.pass.cpp`
lines: 134-135:   
>   test<Context, bool, long>(true, 192812079084L);
     test<Context, bool, long>(false, 192812079084L);
     
 Relands: https://github.com/llvm/llvm-project/pull/76449
 Reverted in: https://github.com/llvm/llvm-project/commit/02f95b77515fe18ed1076b94cbb850ea0cf3c77e

### CODE REVIEW

**Comment by H-G-Hristov** [comment]
The test was failing on chromium CI: https://ci.chromium.org/ui/p/fuchsia/builders/toolchain.ci/clang-windows-x64/b8758295463403595089/test-results?sortby=&groupby=

**Review by mordante — APPROVED** [review]
@petrhosek the original patches worked in our Windows pre-commit CI. Can you verify this patch works for you.

I'm still happy with the patch. Please wait for @petrhosek's confirmation before landing.

**Comment by H-G-Hristov** [comment]
@petrhosek Thank you for reacting. I'm sorry I didn't see your message earlier: https://github.com/llvm/llvm-project/pull/76449#issuecomment-1903063178

**Comment by mstorsjo** [comment]
> @petrhosek the original patches worked in our Windows pre-commit CI.

FWIW, the reason for this is that the test has `UNSUPPORTED: clang-16 || clang-17`, and the Windows pre-commit CI only uses release versions of clang.

**Review by H-G-Hristov — COMMENTED** [review]
> > @petrhosek the original patches worked in our Windows pre-commit CI.
>
> FWIW, the reason for this is that the test has `UNSUPPORTED: clang-16 || clang-17`, and the Windows pre-commit CI only uses release versions of clang. 

Hmm, now that this https://github.com/llvm/llvm-project/pull/78991 has landed we could probably replace the guards with: `__has_extension(cxx_explicit_this_parameter)`

**Comment by petrhosek** [comment]
@H-G-Hristov Would it be possible to update your branch? I sent it to our Windows builders but Clang is segfaulting while building runtimes.

**Comment by H-G-Hristov** [comment]
> @H-G-Hristov Would it be possible to update your branch? I sent it to our Windows builders but Clang is segfaulting while building runtimes.

Thank you for taking a look. I updated the branch. Please note that pre-release Clang 18 used to have a bug related to "deducing this", which is used by this feature. It has been rectified since then: https://github.com/llvm/llvm-project/pull/75913
The other patch related to this paper seems to have passed the CI: https://github.com/llvm/llvm-project/pull/76449

**Comment by H-G-Hristov** [comment]
I noticed some of clang-cl test crashed. I rebased again.

File: libcxx/docs/Status/Cxx2cPapers.csv
Comment (H-G-Hristov) [inline libcxx/docs/Status/Cxx2cPapers.csv:20]:
```suggestion
"`P2637R3 <https://wg21.link/P2637R3>`__","LWG","Member ``visit``","Varna June 2023","|Complete|","19.0",""
```

File: libcxx/docs/Status/FormatIssues.csv
Comment (H-G-Hristov) [inline libcxx/docs/Status/FormatIssues.csv:19]:
```suggestion
"`P2637R3 <https://wg21.link/P2637R3>`__","Member ``visit``","C++26","Hristo Hristov","|Complete|",19.0
```

File: libcxx/docs/ReleaseNotes/18.rst
Comment (H-G-Hristov) [inline libcxx/docs/ReleaseNotes/18.rst:83]:
```suggestion
- P2517R1 - Add a conditional ``noexcept`` specification to ``std::apply``
- P2447R6 - ``span`` over initializer list
```

**Review by H-G-Hristov — COMMENTED** [review]
@mordante I guess the PR missed LLVM18 release. I'll update the release notes accordingly. Is that OK?

**Comment by petrhosek** [comment]
I tried this branch on our Windows builders although I'm not sure if it's related or not to this patch:
```
# COMPILED WITH
C:/b/s/w/ir/x/w/llvm_build/./bin/clang-cl.exe C:\b\s\w\ir\x\w\github-H-G-Hristov-llvm-project\libcxx\test\libcxx\fuzzing\random.pass.cpp --driver-mode=g++ --target=x86_64-pc-windows-msvc -fms-runtime-lib=static -nostdinc++ -I C:/b/s/w/ir/x/w/llvm_build/include/c++/v1 -I C:/b/s/w/ir/x/w/llvm_build/include/x86_64-pc-windows-msvc/c++/v1 -I C:/b/s/w/ir/x/w/github-H-G-Hristov-llvm-project/libcxx/test/support -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_STDIO_ISO_WIDE_SPECIFIERS -DNOMINMAX -std=c++26 -Werror -Wall -Wctad-maybe-unsupported -Wextra -Wshadow -Wundef -Wunused-template -Wno-unused-command-line-argument -Wno-attributes -Wno-pessimizing-move -Wno-noexcept-type -Wno-atomic-alignment -Wno-reserved-module-identifier -Wdeprecated-copy -Wdeprecated-copy-dtor -Wno-user-defined-literals -Wno-tautological-compare -Wsign-compare -Wunused-variable -Wunused-parameter -Wunreachable-code -Wno-unused-local-typedef -Wno-local-type-template-args -Wno-c++11-extensions -Wno-unknown-pragmas -Wno-pass-failed -Wno-mismatched-new-delete -Wno-redundant-move -Wno-self-move -D_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER -D_LIBCPP_ENABLE_EXPERIMENTAL -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_NONE -Werror=thread-safety -Wuser-defined-warnings  -llibc++experimental -nostdlib -L C:/b/s/w/ir/x/w/llvm_build/./lib/x86_64-pc-windows-msvc -llibc++ -llibcpmt -o C:\b\s\w\ir\x\w\llvm_build\runtimes\runtimes-x86_64-pc-windows-msvc-bins\test\libcxx\fuzzing\Output\random.pass.cpp.dir\t.tmp.exe
# executed command: C:/b/s/w/ir/x/w/llvm_build/./bin/clang-cl.exe 'C:\b\s\w\ir\x\w\github-H-G-Hristov-llvm-project\libcxx\test\libcxx\fuzzing\random.pass.cpp' --driver-mode=g++ --target=x86_64-pc-windows-msvc -fms-runtime-lib=static -nostdinc++ -I C:/b/s/w/ir/x/w/llvm_build/include/c++/v1 -I C:/b/s/w/ir/x/w/llvm_build/include/x86_64-pc-windows-msvc/c++/v1 -I C:/b/s/w/ir/x/w/github-H-G-Hristov-llvm-project/libcxx/test/support -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_STDIO_ISO_WIDE_SPECIFIERS -DNOMINMAX -std=c++26 -Werror -Wall -Wctad-maybe-unsupported -Wextra -Wshadow -Wundef -Wunused-template -Wno-unused-command-line-argument -Wno-attributes -Wno-pessimizing-move -Wno-noexcept-type -Wno-atomic-alignment -Wno-reserved-module-identifier -Wdeprecated-copy -Wdeprecated-copy-dtor -Wno-user-defined-literals -Wno-tautological-compare -Wsign-compare -Wunused-variable -Wunused-parameter -Wunreachable-code -Wno-unused-local-typedef -Wno-local-type-template-args -Wno-c++11-extensions -Wno-unknown-pragmas -Wno-pass-failed -Wno-mismatched-new-delete -Wno-redundant-move -Wno-self-move -D_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER -D_LIBCPP_ENABLE_EXPERIMENTAL -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_NONE -Werror=thread-safety -Wuser-defined-warnings -llibc++experimental -nostdlib -L C:/b/s/w/ir/x/w/llvm_build/./lib/x86_64-pc-windows-msvc -llibc++ -llibcpmt -o 'C:\b\s\w\ir\x\w\llvm_build\runtimes\runtimes-x86_64-pc-windows-msvc-bins\test\libcxx\fuzzing\Output\random.pass.cpp.dir\t.tmp.exe'
# .---command stderr------------
# | In file included from C:\b\s\w\ir\x\w\github-H-G-Hristov-llvm-project\libcxx\test\libcxx\fuzzing\random.pass.cpp:16:
# | In file included from C:/b/s/w/ir/x/w/llvm_build/include/c++/v1\cmath:319:
# | In file included from C:/b/s/w/ir/x/w/llvm_build/include/c++/v1\math.h:301:
# | In file included from C:\b\s\w\ir\cache\windows_sdk\Windows Kits\10\Include\10.0.19041.0\ucrt\math.h:11:
# | C:\b\s\w\ir\cache\windows_sdk\Windows Kits\10\Include\10.0.19041.0\ucrt\corecrt_math.h:413:16: error: call to 'fpclassify' is ambiguous
# |   413 |         return fpclassify(_X) == FP_NAN;
# |       |                ^~~~~~~~~~
# | C:\b\s\w\ir\x\w\github-H-G-Hristov-llvm-project\libcxx\test\libcxx\fuzzing\random.pass.cpp:166:12: note: in instantiation of function template specialization 'isnan<short>' requested here
# |   166 |   if (std::isnan(res)) {
# |       |            ^
# | C:\b\s\w\ir\x\w\github-H-G-Hristov-llvm-project\libcxx\test\libcxx\fuzzing\random.pass.cpp:178:10: note: in instantiation of function template specialization 'helper<std::uniform_int_distribution<short>>' requested here
# |   178 |   return helper<std::uniform_int_distribution<std::int16_t>>(data, size)       ||
# |       |          ^
# | C:\b\s\w\ir\cache\windows_sdk\Windows Kits\10\Include\10.0.19041.0\ucrt\corecrt_math.h:288:31: note: candidate function
# |   288 |     _Check_return_ inline int fpclassify(_In_ float _X) throw()
# |       |                               ^
# | C:\b\s\w\ir\cache\windows_sdk\Windows Kits\10\Include\10.0.19041.0\ucrt\corecrt_math.h:293:31: note: candidate function
# |   293 |     _Check_return_ inline int fpclassify(_In_ double _X) throw()
# |       |                               ^
# | C:\b\s\w\ir\cache\windows_sdk\Windows Kits\10\Include\10.0.19041.0\ucrt\corecrt_math.h:298:31: note: candidate function
# |   298 |     _Check_return_ inline int fpclassify(_In_ long double _X) throw()
# |       |                               ^
# | C:\b\s\w\ir\cache\windows_sdk\Windows Kits\10\Include\10.0.19041.0\ucrt\corecrt_math.h:413:16: error: call to 'fpclassify' is ambiguous
# |   413 |         return fpclassify(_X) == FP_NAN;
# |       |                ^~~~~~~~~~
# | C:\b\s\w\ir\x\w\github-H-G-Hristov-llvm-project\libcxx\test\libcxx\fuzzing\random.pass.cpp:166:12: note: in instantiation of function template specialization 'isnan<bool>' requested here
# |   166 |   if (std::isnan(res)) {
# |       |            ^
# | C:\b\s\w\ir\x\w\github-H-G-Hristov-llvm-project\libcxx\test\libcxx\fuzzing\random.pass.cpp:180:10: note: in instantiation of function template specialization 'helper<std::bernoulli_distribution>' requested here
# |   180 |          helper<std::bernoulli_distribution>(data, size)                       ||
# |       |          ^
# | C:\b\s\w\ir\cache\windows_sdk\Windows Kits\10\Include\10.0.19041.0\ucrt\corecrt_math.h:288:31: note: candidate function
# |   288 |     _Check_return_ inline int fpclassify(_In_ float _X) throw()
# |       |                               ^
# | C:\b\s\w\ir\cache\windows_sdk\Windows Kits\10\Include\10.0.19041.0\ucrt\corecrt_math.h:293:31: note: candidate function
# |   293 |     _Check_return_ inline int fpclassify(_In_ double _X) throw()
# |       |                               ^
# | C:\b\s\w\ir\cache\windows_sdk\Windows Kits\10\Include\10.0.19041.0\ucrt\corecrt_math.h:298:31: note: candidate function
# |   298 |     _Check_return_ inline int fpclassify(_In_ long double _X) throw()
# |       |                               ^
# | In file included from C:\b\s\w\ir\x\w\github-H-G-Hristov-llvm-project\libcxx\test\libcxx\fuzzing\random.pass.cpp:20:
# | In file included from C:/b/s/w/ir/x/w/llvm_build/include/c++/v1\random:1686:
# | C:/b/s/w/ir/x/w/llvm_build/include/c++/v1\__random/clamp_to_integral.h:36:86: error: shift count is negative [-Werror,-Wshift-count-negative]
# |    36 |   return _FloatBigger ? numeric_limits<_IntT>::max() : (numeric_limits<_IntT>::max() >> _Bits << _Bits);
# |       |                                                                                      ^  ~~~~~
# | C:/b/s/w/ir/x/w/llvm_build/include/c++/v1\__random/clamp_to_integral.h:46:27: note: in instantiation of function template specialization 'std::__max_representable_int_for_float<short, double, true, -38>' requested here
# |    46 |   const _IntT __max_val = __max_representable_int_for_float<_IntT, _RealT>();
# |       |                           ^
# | C:/b/s/w/ir/x/w/llvm_build/include/c++/v1\__random/poisson_distribution.h:153:21: note: in instantiation of function template specialization 'std::__clamp_to_integral<short, double>' requested here
# |   153 |         return std::__clamp_to_integral<result_type>(__tx);
# |       |                     ^
# | C:/b/s/w/ir/x/w/llvm_build/include/c++/v1\__random/poisson_distribution.h:83:12: note: in instantiation of function template specialization 'std::poisson_distribution<short>::operator()<std::mersenne_twister_engine<unsigned int, 32, 624, 397, 31, 2567483615, 11, 4294967295, 7, 2636928640, 15, 4022730752, 18, 1812433253>>' requested here
# |    83 |     return (*this)(__g, __p_);
# |       |            ^
# | C:\b\s\w\ir\x\w\github-H-G-Hristov-llvm-project\libcxx\test\libcxx\fuzzing\random.pass.cpp:165:23: note: in instantiation of function template specialization 'std::poisson_distribution<short>::operator()<std::mersenne_twister_engine<unsigned int, 32, 624, 397, 31, 2567483615, 11, 4294967295, 7, 2636928640, 15, 4022730752, 18, 1812433253>>' requested here
# |   165 |   volatile auto res = d(engine);
# |       |                       ^
# | C:\b\s\w\ir\x\w\github-H-G-Hristov-llvm-project\libcxx\test\libcxx\fuzzing\random.pass.cpp:181:10: note: in instantiation of function template specialization 'helper<std::poisson_distribution<short>>' requested here
# |   181 |          helper<std::poisson_distribution<std::int16_t>>(data, size)           ||
# |       |          ^
# | In file included from C:\b\s\w\ir\x\w\github-H-G-Hristov-llvm-project\libcxx\test\libcxx\fuzzing\random.pass.cpp:20:
# | In file included from C:/b/s/w/ir/x/w/llvm_build/include/c++/v1\random:1686:
# | C:/b/s/w/ir/x/w/llvm_build/include/c++/v1\__random/clamp_to_integral.h:36:95: error: shift count is negative [-Werror,-Wshift-count-negative]
# |    36 |   return _FloatBigger ? numeric_limits<_IntT>::max() : (numeric_limits<_IntT>::max() >> _Bits << _Bits);
# |       |                                                                                               ^  ~~~~~
# | 4 errors generated.
# `-----------------------------
# error: command failed with exit status: 1

--
```
https://logs.chromium.org/logs/fuchsia/led/phosek_google.com/62644843c966785c9dedf065a79d80df8734b4df5ff7703ea3c2a2b5643cec05/+/u/clang/test/stdout is the full log.

**Comment by Zingam** [comment]
@petrhosek It is not related. It was supposedly fixed by this: https://github.com/llvm/llvm-project/pull/79619
I've been observing the chromium Windows CI failing for a while. Let's see if the above patch fixes it. I'll rebase now this one now.

**Comment by Zingam** [comment]
@petrhosek According to the log, the tests in this patch pass. Could you please try again. I'll wait a while and if there is not negative feedback I'll reland the patch.

> https://logs.chromium.org/logs/fuchsia/led/phosek_google.com/62644843c966785c9dedf065a79d80df8734b4df5ff7703ea3c2a2b5643cec05/+/u/clang/test/stdout is the full log.

```
 PASS: llvm-libc++-static-clangcl.cfg.in :: std/utilities/format/format.arguments/format.arg/visit.pass.cpp (8505 of 9639)
 PASS: llvm-libc++-static-clangcl.cfg.in :: std/utilities/format/format.arguments/format.arg/visit_format_arg.pass.cpp (8506 of 9639)
 PASS: llvm-libc++-static-clangcl.cfg.in :: std/utilities/format/format.arguments/format.arg/visit.return_type.pass.cpp (8537 of 9639)
 PASS: llvm-libc++-static-clangcl.cfg.in :: std/utilities/format/format.arguments/format.arg/visit_format_arg.deprecated.verify.cpp (8439 of 9639)
```

**Comment by H-G-Hristov** [comment]
> I tried this branch on our Windows builders although I'm not sure if it's related or not to this patch:
> 
> ```
> # COMPILED WITH
> C:/b/s/w/ir/x/w/llvm_build/./bin/clang-cl.exe C:\b\s\w\ir\x\w\github-H-G-Hristov-llvm-project\libcxx\test\libcxx\fuzzing\random.pass.cpp --driver-mode=g++ --target=x86_64-pc-windows-msvc -fms-runtime-lib=static -nostdinc++ -I C:/b/s/w/ir/x/w/llvm_build/include/c++/v1 -I C:/b/s/w/ir/x/w/llvm_build/include/x86_64-pc-windows-msvc/c++/v1 -I C:/b/s/w/ir/x/w/github-H-G-Hristov-llvm-project/libcxx/test/support -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_STDIO_ISO_WIDE_SPECIFIERS -DNOMINMAX -std=c++26 -Werror -Wall -Wctad-maybe-unsupported -Wextra -Wshadow -Wundef -Wunused-template -Wno-unused-command-line-argument -Wno-attributes -Wno-pessimizing-move -Wno-noexcept-type -Wno-atomic-alignment -Wno-reserved-module-identifier -Wdeprecated-copy -Wdeprecated-copy-dtor -Wno-user-defined-literals -Wno-tautological-compare -Wsign-compare -Wunused-variable -Wunused-parameter -Wunreachable-code -Wno-unused-local-typedef -Wno-local-type-template-args -Wno-c++11-extensions -Wno-unknown-pragmas -Wno-pass-failed -Wno-mismatched-new-delete -Wno-redundant-move -Wno-self-move -D_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER -D_LIBCPP_ENABLE_EXPERIMENTAL -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_NONE -Werror=thread-safety -Wuser-defined-warnings  -llibc++experimental -nostdlib -L C:/b/s/w/ir/x/w/llvm_build/./lib/x86_64-pc-windows-msvc -llibc++ -llibcpmt -o C:\b\s\w\ir\x\w\llvm_build\runtimes\runtimes-x86_64-pc-windows-msvc-bins\test\libcxx\fuzzing\Output\random.pass.cpp.dir\t.tmp.exe
> # executed command: C:/b/s/w/ir/x/w/llvm_build/./bin/clang-cl.exe 'C:\b\s\w\ir\x\w\github-H-G-Hristov-llvm-project\libcxx\test\libcxx\fuzzing\random.pass.cpp' --driver-mode=g++ --target=x86_64-pc-windows-msvc -fms-runtime-lib=static -nostdinc++ -I C:/b/s/w/ir/x/w/llvm_build/include/c++/v1 -I C:/b/s/w/ir/x/w/llvm_build/include/x86_64-pc-windows-msvc/c++/v1 -I C:/b/s/w/ir/x/w/github-H-G-Hristov-llvm-project/libcxx/test/support -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_STDIO_ISO_WIDE_SPECIFIERS -DNOMINMAX -std=c++26 -Werror -Wall -Wctad-maybe-unsupported -Wextra -Wshadow -Wundef -Wunused-template -Wno-unused-command-line-argument -Wno-attributes -Wno-pessimizing-move -Wno-noexcept-type -Wno-atomic-alignment -Wno-reserved-module-identifier -Wdeprecated-copy -Wdeprecated-copy-dtor -Wno-user-defined-literals -Wno-tautological-compare -Wsign-compare -Wunused-variable -Wunused-parameter -Wunreachable-code -Wno-unused-local-typedef -Wno-local-type-template-args -Wno-c++11-extensions -Wno-unknown-pragmas -Wno-pass-failed -Wno-mismatched-new-delete -Wno-redundant-move -Wno-self-move -D_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER -D_LIBCPP_ENABLE_EXPERIMENTAL -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_NONE -Werror=thread-safety -Wuser-defined-warnings -llibc++experimental -nostdlib -L C:/b/s/w/ir/x/w/llvm_build/./lib/x86_64-pc-windows-msvc -llibc++ -llibcpmt -o 'C:\b\s\w\ir\x\w\llvm_build\runtimes\runtimes-x86_64-pc-windows-msvc-bins\test\libcxx\fuzzing\Output\random.pass.cpp.dir\t.tmp.exe'
> # .---command stderr------------
> # | In file included from C:\b\s\w\ir\x\w\github-H-G-Hristov-llvm-project\libcxx\test\libcxx\fuzzing\random.pass.cpp:16:
> # | In file included from C:/b/s/w/ir/x/w/llvm_build/include/c++/v1\cmath:319:
> # | In file included from C:/b/s/w/ir/x/w/llvm_build/include/c++/v1\math.h:301:
> # | In file included from C:\b\s\w\ir\cache\windows_sdk\Windows Kits\10\Include\10.0.19041.0\ucrt\math.h:11:
> # | C:\b\s\w\ir\cache\windows_sdk\Windows Kits\10\Include\10.0.19041.0\ucrt\corecrt_math.h:413:16: error: call to 'fpclassify' is ambiguous
> # |   413 |         return fpclassify(_X) == FP_NAN;
> # |       |                ^~~~~~~~~~
> # | C:\b\s\w\ir\x\w\github-H-G-Hristov-llvm-project\libcxx\test\libcxx\fuzzing\random.pass.cpp:166:12: note: in instantiation of function template specialization 'isnan<short>' requested here
> # |   166 |   if (std::isnan(res)) {
> # |       |            ^
> # | C:\b\s\w\ir\x\w\github-H-G-Hristov-llvm-project\libcxx\test\libcxx\fuzzing\random.pass.cpp:178:10: note: in instantiation of function template specialization 'helper<std::uniform_int_distribution<short>>' requested here
> # |   178 |   return helper<std::uniform_int_distribution<std::int16_t>>(data, size)       ||
> # |       |          ^
> # | C:\b\s\w\ir\cache\windows_sdk\Windows Kits\10\Include\10.0.19041.0\ucrt\corecrt_math.h:288:31: note: candidate function
> # |   288 |     _Check_return_ inline int fpclassify(_In_ float _X) throw()
> # |       |                               ^
> # | C:\b\s\w\ir\cache\windows_sdk\Windows Kits\10\Include\10.0.19041.0\ucrt\corecrt_math.h:293:31: note: candidate function
> # |   293 |     _Check_return_ inline int fpclassify(_In_ double _X) throw()
> # |       |                               ^
> # | C:\b\s\w\ir\cache\windows_sdk\Windows Kits\10\Include\10.0.19041.0\ucrt\corecrt_math.h:298:31: note: candidate function
> # |   298 |     _Check_return_ inline int fpclassify(_In_ long double _X) throw()
> # |       |                               ^
> # | C:\b\s\w\ir\cache\windows_sdk\Windows Kits\10\Include\10.0.19041.0\ucrt\corecrt_math.h:413:16: error: call to 'fpclassify' is ambiguous
> # |   413 |         return fpclassify(_X) == FP_NAN;
> # |       |                ^~~~~~~~~~
> # | C:\b\s\w\ir\x\w\github-H-G-Hristov-llvm-project\libcxx\test\libcxx\fuzzing\random.pass.cpp:166:12: note: in instantiation of function template specialization 'isnan<bool>' requested here
> # |   166 |   if (std::isnan(res)) {
> # |       |            ^
> # | C:\b\s\w\ir\x\w\github-H-G-Hristov-llvm-project\libcxx\test\libcxx\fuzzing\random.pass.cpp:180:10: note: in instantiation of function template specialization 'helper<std::bernoulli_distribution>' requested here
> # |   180 |          helper<std::bernoulli_distribution>(data, size)                       ||
> # |       |          ^
> # | C:\b\s\w\ir\cache\windows_sdk\Windows Kits\10\Include\10.0.19041.0\ucrt\corecrt_math.h:288:31: note: candidate function
> # |   288 |     _Check_return_ inline int fpclassify(_In_ float _X) throw()
> # |       |                               ^
> # | C:\b\s\w\ir\cache\windows_sdk\Windows Kits\10\Include\10.0.19041.0\ucrt\corecrt_math.h:293:31: note: candidate function
> # |   293 |     _Check_return_ inline int fpclassify(_In_ double _X) throw()
> # |       |                               ^
> # | C:\b\s\w\ir\cache\windows_sdk\Windows Kits\10\Include\10.0.19041.0\ucrt\corecrt_math.h:298:31: note: candidate function
> # |   298 |     _Check_return_ inline int fpclassify(_In_ long double _X) throw()
> # |       |                               ^
> # | In file included from C:\b\s\w\ir\x\w\github-H-G-Hristov-llvm-project\libcxx\test\libcxx\fuzzing\random.pass.cpp:20:
> # | In file included from C:/b/s/w/ir/x/w/llvm_build/include/c++/v1\random:1686:
> # | C:/b/s/w/ir/x/w/llvm_build/include/c++/v1\__random/clamp_to_integral.h:36:86: error: shift count is negative [-Werror,-Wshift-count-negative]
> # |    36 |   return _FloatBigger ? numeric_limits<_IntT>::max() : (numeric_limits<_IntT>::max() >> _Bits << _Bits);
> # |       |                                                                                      ^  ~~~~~
> # | C:/b/s/w/ir/x/w/llvm_build/include/c++/v1\__random/clamp_to_integral.h:46:27: note: in instantiation of function template specialization 'std::__max_representable_int_for_float<short, double, true, -38>' requested here
> # |    46 |   const _IntT __max_val = __max_representable_int_for_float<_IntT, _RealT>();
> # |       |                           ^
> # | C:/b/s/w/ir/x/w/llvm_build/include/c++/v1\__random/poisson_distribution.h:153:21: note: in instantiation of function template specialization 'std::__clamp_to_integral<short, double>' requested here
> # |   153 |         return std::__clamp_to_integral<result_type>(__tx);
> # |       |                     ^
> # | C:/b/s/w/ir/x/w/llvm_build/include/c++/v1\__random/poisson_distribution.h:83:12: note: in instantiation of function template specialization 'std::poisson_distribution<short>::operator()<std::mersenne_twister_engine<unsigned int, 32, 624, 397, 31, 2567483615, 11, 4294967295, 7, 2636928640, 15, 4022730752, 18, 1812433253>>' requested here
> # |    83 |     return (*this)(__g, __p_);
> # |       |            ^
> # | C:\b\s\w\ir\x\w\github-H-G-Hristov-llvm-project\libcxx\test\libcxx\fuzzing\random.pass.cpp:165:23: note: in instantiation of function template specialization 'std::poisson_distribution<short>::operator()<std::mersenne_twister_engine<unsigned int, 32, 624, 397, 31, 2567483615, 11, 4294967295, 7, 2636928640, 15, 4022730752, 18, 1812433253>>' requested here
> # |   165 |   volatile auto res = d(engine);
> # |       |                       ^
> # | C:\b\s\w\ir\x\w\github-H-G-Hristov-llvm-project\libcxx\test\libcxx\fuzzing\random.pass.cpp:181:10: note: in instantiation of function template specialization 'helper<std::poisson_distribution<short>>' requested here
> # |   181 |          helper<std::poisson_distribution<std::int16_t>>(data, size)           ||
> # |       |          ^
> # | In file included from C:\b\s\w\ir\x\w\github-H-G-Hristov-llvm-project\libcxx\test\libcxx\fuzzing\random.pass.cpp:20:
> # | In file included from C:/b/s/w/ir/x/w/llvm_build/include/c++/v1\random:1686:
> # | C:/b/s/w/ir/x/w/llvm_build/include/c++/v1\__random/clamp_to_integral.h:36:95: error: shift count is negative [-Werror,-Wshift-count-negative]
> # |    36 |   return _FloatBigger ? numeric_limits<_IntT>::max() : (numeric_limits<_IntT>::max() >> _Bits << _Bits);
> # |       |                                                                                               ^  ~~~~~
> # | 4 errors generated.
> # `-----------------------------
> # error: command failed with exit status: 1
> 
> --
> ```
> 
> https://logs.chromium.org/logs/fuchsia/led/phosek_google.com/62644843c966785c9dedf065a79d80df8734b4df5ff7703ea3c2a2b5643cec05/+/u/clang/test/stdout is the full log.

This issue was already reported a while back here: https://github.com/llvm/llvm-project/issues/70225

**Comment by mordante** [comment]
> @mordante I guess the PR missed LLVM18 release. I'll update the release notes accordingly. Is that OK?

Yes. It would have been nice to get this in LLVM 18, but I don't feel it's critical to backport it. (Especially since it's a C++26 only feature and I expect very few people are already using it.)

**Pull request merged**
